core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -82,48 +99,152 @@ public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardR
+            Set<String> inSyncAllocationIds = new HashSet<>(oldInSyncAllocationIds);
+            inSyncAllocationIds.addAll(updates.addedAllocationIds);
+            inSyncAllocationIds.removeAll(updates.removedAllocationIds);
+
+            // Prevent set of inSyncAllocationIds to grow unboundedly. This can happen for example if we don't write to a primary
+            // but repeatedly shut down nodes that have active replicas.
+            // We use number_of_replicas + 1 (= possible active shard copies) to bound the inSyncAllocationIds set
+            int upperBound = oldIndexMetaData.getNumberOfReplicas() + 1; // +1 for the primary
+            int overflow = inSyncAllocationIds.size() - upperBound;
+            if (overflow > 0) {
+                // trim entries that have no corresponding shard routing in the cluster state (i.e. trim unavailable copies)
+                Set<String> assignedAllocations = newRoutingTable.shardRoutingTable(shardId).assignedShards().stream()
+                    .map(shr -> shr.allocationId().getId()).collect(Collectors.toSet());
+                Set<String> candidatesToRemove = Sets.difference(inSyncAllocationIds, assignedAllocations);
+                for (Iterator<String> it = candidatesToRemove.iterator(); it.hasNext() && overflow > 0; ) {
+                    inSyncAllocationIds.remove(it.next());
+                    overflow--;
+                }
+                assert overflow == 0 :
+                    "not enough candidates to remove, candidates: " + candidatesToRemove + " overflow: " + overflow;
+            }
+
+            // only update in-sync allocation ids if there is at least one entry remaining. Assume for example that there only
+            // ever was a primary active and now it failed. If we were to remove the allocation id from the in-sync set, this would
+            // create an empty primary on the next allocation (see ShardRouting#allocatedPostIndexCreate)
+            if (inSyncAllocationIds.isEmpty() && oldInSyncAllocationIds.isEmpty() == false) {
+                assert updates.firstFailedPrimary != null :
+                    "in-sync set became empty but active primary wasn't failed: " + oldInSyncAllocationIds;
+                if (updates.firstFailedPrimary != null) {
+                    // add back allocation id of failed primary
+                    inSyncAllocationIds.add(updates.firstFailedPrimary.allocationId().getId());
+                }
+            }
+
+            // be extra safe here and only update in-sync set if it is non-empty
+            if (inSyncAllocationIds.isEmpty() == false) {
+                assert inSyncAllocationIds.isEmpty() == false;
+                if (indexMetaDataBuilder == null) {
+                    indexMetaDataBuilder = IndexMetaData.builder(oldIndexMetaData);
+                }
+                indexMetaDataBuilder.putInSyncAllocationIds(shardId.id(), inSyncAllocationIds);
+            }
```

Comment:
I feel like we should log a warning here to alert to the fact that we have an empty in-sync allocation id set.  Even though we didn't update the set in the index meta data, the fact that we could get to this position signals a problem, especially since we ignore if `firstFailedPrimary` is null here: https://github.com/elastic/elasticsearch/pull/20023/files#diff-ff816fc88b23166c4be9374415818aceR202.  Either we should not have the conditional